libtenzir/builtins/operators/load_balance.cpp

```diff
@@ -0,0 +1,396 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "tenzir/pipeline_executor.hpp"
+#include "tenzir/scope_linked.hpp"
+#include "tenzir/tql2/exec.hpp"
+
+#include <tenzir/tql2/plugin.hpp>
+
+#include <deque>
+
+namespace tenzir::plugins::load_balance {
+
+namespace {
+
+using load_balancer_actor = caf::typed_actor<
+  // Write events to a consumer pipeline, waiting for a read.
+  auto(atom::write, table_slice events)->caf::result<void>,
+  // Read events, waiting for a write.
+  auto(atom::read)->caf::result<table_slice>>
+  // Handle metrics of the nested pipelines.
+  ::extend_with<metrics_receiver_actor>
+  // Handle diagnostics of the nested pipelines.
+  ::extend_with<receiver_actor<diagnostic>>;
+
+/// Load balancing is currently done in a naive, yet hopefully effective way:
+/// - Every subpipeline configuration gets its own executor.
+/// - When the source of that pipeline is pulled from, it requests a batch from
+///   the load balancer.
+/// - The input of the `load_balance` operator is sent to the load balancer as
+///   well in order to be forwarded to these read requests.
+/// - Read requests are fulfilled in a FIFO manner, guaranteeing some degree of
+///   fairness.
+/// - The write operation only returns when the batch has been read from.
+/// - Because of the implicit buffering between operators, upstream can still
+///   continue even if the write is being blocked.
+/// - We currently hand out batches exactly as they come in. Thus, their size
+///   can vary significantly, producing an uneven load across the instances.
+/// If our handover strategy here unexpectedly turns out to be a bottleneck,
+/// then it should not be too hard to switch to a different mechanism.
+struct load_balancer_state {
+  [[maybe_unused]] static constexpr auto name = "load-balancer";
+
+  load_balancer_actor::pointer self{};
+  shared_diagnostic_handler diagnostics;
+  metrics_receiver_actor metrics;
+  std::vector<pipeline_executor_actor> executors;
+  std::deque<caf::typed_response_promise<table_slice>> reads;
+  std::deque<std::pair<table_slice, caf::typed_response_promise<void>>> writes;
+  bool finished = false;
+  uint64_t operator_index{};
+  uint64_t next_metrics_id = 0;
+  detail::stable_map<std::pair<uint64_t, uint64_t>, uint64_t> metrics_id_map;
+
+  auto write(table_slice events) -> caf::result<void> {
+    TENZIR_ASSERT(events.rows() > 0);
+    if (not reads.empty()) {
+      TENZIR_WARN("writing {} events directly", events.rows());
```

Comment:
```suggestion
```

There's also a bunch of other leftover comments from development or debugging, please remove them as well.

Reply:
My bad, I demoted them previously to `VERBOSE` but then accidentally committed a version with `WARN` again. Demoted them again. Might be useful to keep them for a bit in case hard-to-reproduce problems with this come up. But I can also fully remove them if that's what you'd prefer.

Reply:
`VERBOSE` in a hot loop is still too noisy. @tobim just demoted a whole bunch of log messages to `TRACE` or removed them entirely because they flooded production log files.

Reply:
Which leftover comments are you talking about? The comments all look good to me.

Reply:
`s/leftover comments/leftover logs`, whoops 🙈

Reply:
> VERBOSE in a hot loop is still too noisy. 

Makes sense. I'll remove all log messages in the inner loops then, and demote the remaining lifecycle-related messages to `DEBUG`. Please let me know if that's not good enough.

Reply:
That should do the trick.